### PR TITLE
Schema validation

### PR DESCRIPTION
**Pull request by cptartur**
TODO 

- [x] disabling validation for requests and responses
- [x] handling optional request parameters
- [ ] create an issue in devnet repo about getEvents response being incorrect
- [ ] consider alternative methods of loading schemas (from file/from git submodule)
- [x] manually test disabling response validation

### CODE REVIEW

File: test/support/assertions.py [hunk 1]
```diff
@@ -9,13 +9,13 @@
...
 
-    schema = _load_json_schema(schema_file)
+    schema = load_json_schema(schema_file)
     schema_dir = join(dirname(__file__), "schemas")
     resolver = RefResolver(base_uri="file://" + schema_dir + "/", referrer=schema)
     return validate(data, schema, resolver=resolver)
 
 
-def _load_json_schema(filename):
+def load_json_schema(filename):
```
Comment (war-in) [inline test/support/assertions.py]: Maybe other functions should be moved here? (`assert_valid_rpc_schema`)?

File: test/rpc/schema.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+# Cache the function result so schemas are not reloaded from disk on every call
+@lru_cache
+def _load_schemas() -> Tuple[Dict[str, Any], Dict[str, Any]]:
+    specs_json = load_json_schema("starknet_api_openrpc.json")
+    schemas = specs_json["components"]["schemas"]
+    methods = {method["name"]: method for method in specs_json["methods"]}
+
+    for schema in schemas.values():
+        # Newer version of the rpc (above 0.45.0) have required properly defined.
```
Comment (war-in) [inline test/rpc/schema.py]:
```suggestion
        # Newer version of the RPC (above 0.45.0) has properly defined `required` fields.
```

File: test/rpc/schema.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            schema["required"] = [prop for prop in schema["properties"].keys()]
+
+        # Ensures validation fails in case of extra fields not matched by any of `allOf` / `anyOf` branches.
+        if "allOf" in schema or "anyOf" in schema:
+            schema["unevaluatedProperties"] = False
+
+    return methods, schemas
+
+
+def assert_valid_rpc_schema(data: Dict[str, Any], method_name: str):
```
Comment (war-in) [inline test/rpc/schema.py]: Could be placed after `_schema_for_method` function (all above them all)

File: test/rpc/schema.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+from functools import lru_cache
+
+from test.support.assertions import load_json_schema
+from typing import Any, Dict, Tuple
+
+from jsonschema import validate
+
+
+# Cache the function result so schemas are not reloaded from disk on every call
+@lru_cache
```
Comment (war-in) [inline test/rpc/schema.py]: 🔥

File: test/support/schemas/starknet_api_openrpc.json [hunk 5]
```diff
@@ -0,0 +1,1964 @@
+{
```
Comment (war-in) [inline test/support/schemas/starknet_api_openrpc.json]: What about methods from the [write api](https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_write_api.json)?

Comment (war-in) [reply test/support/schemas/starknet_api_openrpc.json]: I'm not sure about keeping such files in the repo. Maybe there is another way of achieving that? Or do you think it is ok?

**Review by war-in — COMMENTED** [review]
Could you show a screenshot or something showing an error when something is wrong?

File: test/support/schemas/starknet_api_openrpc.json (hunk 5 above)
Comment (cptartur) [reply test/support/schemas/starknet_api_openrpc.json]: I don't like keeping it in the repo either. This is a modified version of the RPC spec (some usages of `oneOf` replaced with `anyOf`, fixed some bugs), because the 0.2.1 release had some problems that interfered with the validation.

Ideally (once new rpc spec is released hopefully addressing this problem), we could use a git submodule instead.

Comment (war-in) [reply test/support/schemas/starknet_api_openrpc.json]: Yup, it will be the best solution. We should keep it in mind

File: test/support/assertions.py (hunk 1 above)
Comment (cptartur) [reply test/support/assertions.py]: Not sure about this. Methods added by me are very much RPC specific.

**Comment by cptartur** [comment]
> Could you show a screenshot or something showing an error when something is wrong?

Sure, here's an example error from test_get_events

image

**Comment by war-in** [comment]
> > Could you show a screenshot or something showing an error when something is wrong?
> 
> Sure, here's an example error from test_get_events
> 
> image

Does it show which test is failing? I assume so but would like o be sure 😅

File: test/support/assertions.py (hunk 1 above)
Comment (war-in) [reply test/support/assertions.py]: My point is that `load_json_schema` was private. But you're right they are connected only to the RPC

**Comment by cptartur** [comment]
> > > Could you show a screenshot or something showing an error when something is wrong?
> > 
> > 
> > Sure, here's an example error from test_get_events
> > image
> 
> Does it show which test is failing? I assume so but would like o be sure 😅

Yeah, it does. I cropped the rest of the output.

File: starknet_devnet/blueprints/rpc/schema.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    """
+    Wrapper for ValidationError raised during response validation
+    """
+
+    def __init__(self, err: ValidationError):
+        super().__init__()
+        self.validation_error = err
+
+
+def require_valid_request_and_response(method_name: str):
```
Comment (Solpatium) [inline starknet_devnet/blueprints/rpc/schema.py]: validate_schema?

File: starknet_devnet/blueprints/rpc/schema.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        validate(value, schemas.schemas[name])
+
+
+class ParamsValidationErrorWrapper(Exception):
+    """
+    Wrapper for ValidationError raised during request validation
+    """
+
+    def __init__(self, err: ValidationError):
+        super().__init__()
```
Comment (Solpatium) [inline starknet_devnet/blueprints/rpc/schema.py]: Message should be provided to parent, same below.

File: starknet_devnet/blueprints/rpc/schema.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+
+@dataclass
+class RequestSchema:
+    """
+    Return type of _request_schemas_for_method function
+    """
+
+    params: List[str]
+    schemas: Dict[str, Any]
```
Comment (Solpatium) [inline starknet_devnet/blueprints/rpc/schema.py]: Why do you need params here as well? It seems that schemas have their names as keys. Python 3.7+ requieres order, but to be explicit you can also use OrderedDict.

File: starknet_devnet/blueprints/rpc/schema.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+def _response_schema_for_method(name: str) -> Dict[str, Any]:
+    """
+    Return a dict structured like
+    {
+        // base schema
+        // ...
+        "components": {
+            "schemas": {
+                // rest of the schemas
```
Comment (Solpatium) [inline starknet_devnet/blueprints/rpc/schema.py:54]: So those are schemas for fields that might be used in the base schema, right? I think we should include this info.

File: starknet_devnet/blueprints/rpc/schema.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+            }
+        }
+    }
+    for the use in json schema validation.
+    """
+
+    methods, schemas = _load_schemas()
+    base_schema = methods[name]["result"]["schema"]
+
+    if all(i not in base_schema for i in ("allOf", "anyOf", "oneOf")):
```
Comment (Solpatium) [inline starknet_devnet/blueprints/rpc/schema.py]: I think `not any(rule in base_schema for rule in ("allOf", "anyOf", "oneOf"))` might be more readable

File: starknet_devnet/blueprints/rpc/schema.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        schema = {**param["schema"], "components": {"schemas": schemas}}
+        request_schemas[name] = schema
+
+    return RequestSchema(params, request_schemas)
+
+
+def _construct_method_name(method_name: str) -> str:
+    if "starknet_" not in method_name:
+        method_name = "starknet_" + method_name
+    return method_name
```
Comment (Solpatium) [inline starknet_devnet/blueprints/rpc/schema.py]: I would avoid doing this conditionally.

File: starknet_devnet/blueprints/rpc/schema.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+def _load_schemas() -> Tuple[Dict[str, Any], Dict[str, Any]]:
+    specs_json = json.loads(RPC_SPECIFICATION)
+    schemas = specs_json["components"]["schemas"]
+    methods = {method["name"]: method for method in specs_json["methods"]}
+
+    write_specs_json = json.loads(RPC_SPECIFICATION_WRITE)
+    methods = {
+        **methods,
+        **{method["name"]: method for method in write_specs_json["methods"]},
+    }
```
Comment (Solpatium) [inline starknet_devnet/blueprints/rpc/schema.py]: Joining methods should be simplified

File: starknet_devnet/blueprints/rpc/schema.py (hunk 11 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/schema.py]: I want to support passing rpc method names both as eg. `starknet_call` and `call`. What's the problem with this method?

File: starknet_devnet/blueprints/rpc/schema.py (hunk 8 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/schema.py]: I think it can be replaced with just an OrderedDict. I'll try adding some tests though, to make sure it works correctly.

File: starknet_devnet/blueprints/rpc/schema.py (hunk 12 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/schema.py]: Not exactly sure how, but I'll try making it simpler.

Comment (Solpatium) [reply starknet_devnet/blueprints/rpc/schema.py]: You can merge them together first and then apply the `{method["name"]: method for method in write_specs_json["methods"]}` mapping.

File: starknet_devnet/blueprints/rpc/schema.py (hunk 11 above)
Comment (Solpatium) [reply starknet_devnet/blueprints/rpc/schema.py]: It is not a problem, just a nitpick. 
What's the benefit here? People will just use short versions, so why not always add the prefix? It is not like some public users will use it and we have to cover both cases.

File: starknet_devnet/blueprints/rpc/schema.py (hunk 12 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/schema.py]: I added a helper function and merge results of it. I think it's quite readable.

**Comment by cptartur** [comment]
Logic is I think done. What I want to add is option to disable validation separately for requests and responses. I can imagine the situation where (especially with responses), a bug is uncovered after merge and it messes with someone tests by throwing incorrect validation errors.

File: starknet_devnet/blueprints/rpc/structures/responses.py [hunk 13]
```diff
@@ -176,15 +175,11 @@ def events() -> List[Event]:
...
-            return "UNKNOWN"
-
-        mapping = {
-            TransactionStatus.NOT_RECEIVED: "UNKNOWN",
+    def status() -> TxnStatus:
+        mapping: Dict[TransactionStatus, TxnStatus] = {
             TransactionStatus.ACCEPTED_ON_L2: "ACCEPTED_ON_L2",
             TransactionStatus.ACCEPTED_ON_L1: "ACCEPTED_ON_L1",
-            TransactionStatus.RECEIVED: "RECEIVED",
+            TransactionStatus.RECEIVED: "PENDING",
```
Comment (war-in) [inline starknet_devnet/blueprints/rpc/structures/responses.py:182]: Hmm, I do not like this one :/ Does spec define it this way?

File: starknet_devnet/blueprints/rpc/structures/responses.py [hunk 14]
```diff
@@ -200,8 +195,10 @@ async def txn_type() -> RpcTxnType:
         "transaction_hash": rpc_felt(txr.transaction_hash),
         "actual_fee": rpc_felt(txr.actual_fee or 0),
         "status": status(),
-        "block_hash": rpc_felt(txr.block_hash) if txr.block_hash is not None else None,
-        "block_number": txr.block_number or None,
+        "block_hash": rpc_felt(txr.block_hash)
+        if txr.block_hash is not None
+        else rpc_felt(0),
+        "block_number": txr.block_number if txr.block_number is not None else 0,
```
Comment (war-in) [inline starknet_devnet/blueprints/rpc/structures/responses.py:201]: Should it be set to `0` as a default?

File: test/rpc/test_rpc_call.py [hunk 15]
```diff
@@ -23,14 +23,15 @@ def test_call(deploy_info, latest_block_id):
...
-                "entry_point_selector": hex(get_selector_from_name("get_balance")),
+                "entry_point_selector": rpc_felt(get_selector_from_name("get_balance")),
                 "calldata": [],
             },
             "block_id": latest_block_id,
         },
     )
     assert "error" not in resp
     result = resp["result"]
+    assert_valid_rpc_schema(result, "starknet_call")
```
Comment (war-in) [inline test/rpc/test_rpc_call.py]: Isn't it validated by default? (by decorator I mean)

File: test/rpc/test_rpc_misc.py [hunk 16]
```diff
@@ -148,36 +154,37 @@ def test_call_with_invalid_params(params):
...
+#         assert str(resp["result"]["events"][i]["data"]) == str(data)
+#
+#     if "continuation_token" in input_data:
+#         expected_continuation_token = int(input_data["continuation_token"])
+#
+#         # increase continuation_token when events are not empty
+#         if resp["result"]["events"]:
+#             expected_continuation_token += 1
+#
+#         assert expected_continuation_token == int(resp["result"]["continuation_token"])
```
Comment (war-in) [inline test/rpc/test_rpc_misc.py]: 👀

File: test/rpc/test_rpc_schema.py [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+def test_schema_raises_on_invalid_kwargs(params):
+    """
+    Call contract
+    """
+    resp = rpc_call("starknet_call", params=params)
+
+    assert "error" in resp
+    error = resp["error"]
+
+    print(error)
```
Comment (war-in) [inline test/rpc/test_rpc_schema.py]: 👀

File: test/rpc/test_rpc_schema.py [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+            },
+            "block_id": "latest",
+        },
+        {
+            "request": {
+                "contract_address": "0x01",
+                "entry_point_selector": ["0x01", "0x02"],
+                "calldata": [],
+            },
+            "block_id": "latest",
```
Comment (war-in) [inline test/rpc/test_rpc_schema.py:75]: What about the test with `block_id` as a number? Is it makes sense?

File: test/rpc/test_rpc_schema.py [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+def test_schema_does_not_raise_on_correct_kwargs(params):
+    """
+    Call contract
+    """
+
+    resp = rpc_call("starknet_call", params=params)
+
+    # Error will be raised because address is correctly formatted but incorrect
+    error = resp["error"]
+    print(error)
```
Comment (war-in) [inline test/rpc/test_rpc_schema.py]: 👀

File: starknet_devnet/blueprints/rpc/structures/responses.py (hunk 13 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/structures/responses.py:182]: RPC 0.2.1 spec only defines these
```json
"TXN_STATUS": {
    "type": "string",
    "enum": [
        "PENDING",
        "ACCEPTED_ON_L2",
        "ACCEPTED_ON_L1",
        "REJECTED"
    ],
    "description": "The status of the transaction"
}
```

File: starknet_devnet/blueprints/rpc/structures/responses.py (hunk 14 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/structures/responses.py:201]: It should have some value (rpc spec from main branch even defines block_hash and block_number as required parameters) and I though 0 is a reasonable default since it has to be a non negative value.

File: test/rpc/test_rpc_call.py (hunk 15 above)
Comment (cptartur) [reply test/rpc/test_rpc_call.py]: Good catch, this is a leftover of when validation was supposed to be only done in tests. I'll remove that.

File: test/rpc/test_rpc_misc.py (hunk 16 above)
Comment (cptartur) [reply test/rpc/test_rpc_misc.py]: Also a good catch. There's some kind of problem with events endpoint spec compliance. I'll probably restore these and address the error in the separate PR.

File: test/rpc/test_rpc_schema.py (hunk 18 above)
Comment (cptartur) [reply test/rpc/test_rpc_schema.py:75]: It does. I'll add some.

File: starknet_devnet/blueprints/rpc/schema.py [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+        return
+
+    if kwargs:
+        if len(kwargs) > len(schemas):
+            raise ValidationError("Too many arguments provided.")
+
+        for name, schema in schemas.items():
+            if name not in kwargs:
+                if schema["is_required"]:
+                    raise ValidationError(f"Missing kwarg for {name}.")
```
Comment (cptartur) [inline starknet_devnet/blueprints/rpc/schema.py]: This is untested: there are no methods with optional parameters in the rpc currently.

File: starknet_devnet/blueprints/rpc/misc.py [hunk 21]
```diff
@@ -64,6 +67,9 @@ async def syncing() -> Union[dict, bool]:
 
 
 # pylint: disable=too-many-arguments
+# Events response does not currently conform to RPC specs
+# and will need fixing before validation is added
+# @validate_schema("getEvents")
 async def get_events(
```
Comment (cptartur) [inline starknet_devnet/blueprints/rpc/misc.py:73]: This will have to be addressed in a separate PR

**Review by cptartur — COMMENTED** [review]
I'm not sure how to make tests for response validation like I did for requests. It is only being tested by running methods calling it. Same for disabling response validation.

Only idea I have is creating mock rpc routes but that is not a good solution in my opinion.

File: starknet_devnet/blueprints/rpc/structures/responses.py (hunk 14 above)
Comment (war-in) [reply starknet_devnet/blueprints/rpc/structures/responses.py:201]: Don't you think it should be set to "latest"?

**Comment by Solpatium** [comment]
>I'm not sure how to make tests for response validation like I did for requests. It is only being tested by running methods calling it. Same for disabling response validation.

You have to test validation itself, so testing it by creating a function decorated with your decorator that returns something is fine.

File: starknet_devnet/blueprints/rpc/misc.py (hunk 21 above)
Comment (Solpatium) [reply starknet_devnet/blueprints/rpc/misc.py:73]: This is the way. Not everything has to be ready from the start.

File: starknet_devnet/blueprints/rpc/schema.py (hunk 20 above)
Comment (war-in) [reply starknet_devnet/blueprints/rpc/schema.py]: How about creating a `fake` RPC schema to test it? 😏

File: starknet_devnet/blueprints/rpc/misc.py (hunk 21 above)
Comment (war-in) [reply starknet_devnet/blueprints/rpc/misc.py:73]: Yup, we should add an issue in the `devnet` repo after PR is published

File: starknet_devnet/blueprints/rpc/routes.py [hunk 22]
```diff
@@ -128,6 +128,11 @@ def parse_body(body: dict) -> Tuple[Callable, Union[List, dict], int]:
         )
 
     if not isinstance(params, (List, Dict)):
-        raise RpcError(code=RpcErrorCode.INVALID_PARAMS.value, message="Invalid params")
+        raise RpcError(
+            code=RpcErrorCode.INVALID_PARAMS.value,
+            # fmt: off
+            message="Invalid \"params\" type. Value of \"params\" must be a dict or list",
+            # fmt: on
```
Comment (war-in) [inline starknet_devnet/blueprints/rpc/routes.py:135]: Why the formatting is disabled? :/

File: starknet_devnet/blueprints/rpc/schema.py [hunk 23]
```diff
@@ -147,7 +150,9 @@ def assert_valid_rpc_request(*args, method_name: str, **kwargs):
 
         for name, schema in schemas.items():
             if name not in kwargs:
-                raise ValidationError(f"Missing kwarg for {name}.")
+                if schema["is_required"]:
+                    raise ValidationError(f"Missing kwarg for {name}.")
+                continue
```
Comment (war-in) [inline starknet_devnet/blueprints/rpc/schema.py:164]: Do we need `continue` here?

File: starknet_devnet/blueprints/rpc/schema.py [hunk 24]
```diff
@@ -169,7 +174,9 @@ def __init__(self, err: ValidationError):
         self.validation_error = err
 
     def __str__(self):
-        return f'Got invalid value for parameter: "{self.validation_error.message}"'
+        # fmt: off
+        return f"Got invalid value for parameter: \"{self.validation_error.message}\""
+        # fmt: on
```
Comment (war-in) [inline starknet_devnet/blueprints/rpc/schema.py:188]: I do not like disabling `black` anywhere

File: starknet_devnet/blueprints/rpc/routes.py (hunk 22 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/routes.py:135]: For whatever reason black doesn't like escaped double quotes, so it converts the message to `'Invalid "params" type...'` but then pylint complains that invalid quotes (`'` instead of `"`) are being used.

So it's either disabling autoformatting or pylint here

File: starknet_devnet/blueprints/rpc/schema.py (hunk 23 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/schema.py:164]: If there's no value for kwarg and it is not required, we should just skip the value and continue validating next kwarg.

File: starknet_devnet/blueprints/rpc/schema.py (hunk 20 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/schema.py]: Hmm, this could be done with some patching I think. I'll test that.

File: starknet_devnet/blueprints/rpc/schema.py (hunk 24 above)
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/schema.py:188]: See my response above

**Review by war-in — APPROVED** [review]

**Comment by cptartur** [comment]
I manually verified if ` --disable-rpc-response-validation` works by changing the response of AddDeclareTransaction to an incorrect one, and it does. RPC returned the response without raising an exception.
